pkg/cvo/upgradeable.go

```diff
@@ -0,0 +1,215 @@
+
+// syncUpgradeable. The status is only checked if it has been more than
+// the minimumUpdateCheckInterval since the last check.
+func (optr *Operator) syncUpgradeable(config *configv1.ClusterVersion) error {
+	// updates are only checked at most once per minimumUpdateCheckInterval or if the generation changes
+	u := optr.getUpgradeable()
+	if u != nil && u.RecentlyChanged(optr.minimumUpdateCheckInterval) {
+		klog.V(4).Infof("Upgradeable conditions were recently checked, will try later.")
+		return nil
+	}
+
+	now := metav1.Now()
+	var conds []configv1.ClusterOperatorStatusCondition
+	var reasons []string
+	for _, check := range optr.upgradeableChecks {
+		if cond := check.Check(); cond != nil {
+			reasons = append(reasons, cond.Reason)
+			cond.LastTransitionTime = now
+			conds = append(conds, *cond)
+		}
+	}
+	if len(conds) == 1 {
+		conds = append(conds, configv1.ClusterOperatorStatusCondition{
+			Type:               configv1.OperatorUpgradeable,
+			Status:             configv1.ConditionFalse,
+			Reason:             conds[0].Reason,
+			Message:            conds[0].Message,
+			LastTransitionTime: now,
+		})
+	} else if len(conds) > 1 {
+		conds = append(conds, configv1.ClusterOperatorStatusCondition{
+			Type:               configv1.OperatorUpgradeable,
+			Status:             configv1.ConditionFalse,
+			Reason:             "MultipleReasons",
+			Message:            fmt.Sprintf("Cluster is not Upgradeable due to multiple reasons: %s", strings.Join(reasons, ",")),
+			LastTransitionTime: now,
+		})
+	}
+	sort.Slice(conds, func(i, j int) bool { return conds[i].Type < conds[j].Type })
+	optr.setUpgradeable(&upgradeable{
+		Conditions: conds,
+	})
+	// requeue
+	optr.queue.Add(optr.queueKey())
+	return nil
+}
+
+type upgradeable struct {
+	At time.Time
+
+	// these are sorted by Type
+	Conditions []configv1.ClusterOperatorStatusCondition
+}
+
+func (u *upgradeable) RecentlyChanged(interval time.Duration) bool {
+	return u.At.After(time.Now().Add(-interval))
+}
+
+func (u *upgradeable) NeedsUpdate(original *configv1.ClusterVersion) *configv1.ClusterVersion {
+	if u == nil {
+		return nil
+	}
+
+	origUpConditions := collectUpgradeableConditions(original.Status.Conditions)
+	if equality.Semantic.DeepEqual(u.Conditions, origUpConditions) {
+		return nil
+	}
+
+	config := original.DeepCopy()
+	for _, c := range u.Conditions {
+		resourcemerge.SetOperatorStatusCondition(&config.Status.Conditions, c)
+	}
+	for _, origc := range origUpConditions {
+		if c := resourcemerge.FindOperatorStatusCondition(u.Conditions, origc.Type); c == nil {
+			resourcemerge.RemoveOperatorStatusCondition(&config.Status.Conditions, origc.Type)
+		}
+	}
+	return config
+}
+
+func collectUpgradeableConditions(conditions []configv1.ClusterOperatorStatusCondition) []configv1.ClusterOperatorStatusCondition {
+	var ret []configv1.ClusterOperatorStatusCondition
+	for _, c := range conditions {
+		if strings.HasPrefix(string(c.Type), string(configv1.OperatorUpgradeable)) {
```

Comment:
Hrm, is this really a good idea?  Do we *need* multiple nested condition types?

Reply:
_my thought was_
It gives us the opportunity to a provide detailed message for different cases why upgradeable is false, but still allow `Upgradeable` condition to be summarize if only one type of failure exists, and link to individual ones when there are multiple types of reasons.. like https://github.com/openshift/cluster-version-operator/pull/243/files#diff-745b27730a4db8ced457d29d6656b64aR3022-R3037

do you think single condition is more cleaner and enough?

Reply:
also the multiple conditions only happen when there are more than one type of failure... for most cases it will be single condition.